### PR TITLE
Update 1.12.1

### PR DESCRIPTION
**Pull request by danpetry**

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -1,8 +1,10 @@
-#!/bin/bash -e
+#!/bin/bash
```
Comment (katietz) [inline recipe/build.sh:1]: agreed to this change, but you should be aware that the she-bang isn't used during conda-build anyway, as it rewrites the scripts ....

File: recipe/build.sh [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
 
-echo "####################################################################" >&2
-echo "Building PyTorch using BLAS implementation: $blas_impl              " >&2
-echo "####################################################################" >&2
+set -x
```
Comment (katietz) [inline recipe/build.sh:3]: shouldn't this be '-ex' ?

File: recipe/build.sh [hunk 3]
```diff
@@ -64,28 +66,22 @@ case "$build_platform" in
...
 
 # MacOS build is simple, and will not be done for CUDA.
 if [[ "$OSTYPE" == "darwin"* ]]; then
-    # When running the full test suite, some tests expect to find resources in
-    # work and work/build, hence the in-tree build.
     "$PYTHON" -m pip install . \
         --no-deps \
         --no-binary :all: \
         --no-clean \
-        --use-feature=in-tree-build \
```
Comment (katietz) [inline recipe/build.sh:88]: is this in-tree feature causing issues here?

File: recipe/meta.yaml [hunk 4]
```diff
@@ -108,13 +114,11 @@ requirements:
...
-    # The test suite uses features only available in numpy 1.17+. We override
-    # numpy pinnings here due to a bug in conda-build mishandling zip key
-    # overrides in local cbc.yaml files. Remove and/or adjust this on the next
-    # upgrade.
-    - numpy >=1.17                    # [not (osx and arm64) and py<310 ]
-    - numpy                           # [    (osx and arm64) or  py>=310]
-    - pip
+    # Pinnings as per pytorch/.circleci/docker/common/install_conda.sh,
+    # except for python 3.8 because we don't have the numpy version required
+    - numpy =1.21.2 # [py>=310]
```
Comment (katietz) [inline recipe/meta.yaml:119]: IMHO we should pin here without assignment.  for py>=310 version '>=1.21.2,<1.22.0' should be better

File: recipe/meta.yaml [hunk 5]
```diff
@@ -108,13 +114,11 @@ requirements:
...
+    - numpy =1.19.2 # [py<=39]
```
Comment (katietz) [inline recipe/meta.yaml:120]: and here '>=1.19.2,<1.21.0'

File: recipe/meta.yaml [hunk 6]
```diff
@@ -142,8 +146,8 @@ requirements:
     - {{ pin_compatible('numpy') }}
```
Comment (katietz) [inline recipe/meta.yaml:146]: replace this with '>=1.21.2,<2' for py>=310 and otherwise for '>=1.19.2,<2'

File: recipe/build.sh (hunk 1 above)
Comment (danpetry) [reply recipe/build.sh:1]: Ah, that must be why `set` was working but editing the flags here wasn't. This change was just a side-effect from some test changes.

Comment (danpetry) [reply recipe/build.sh:1]: I've put `set -ex`, it should exit if any line fails

File: recipe/build.sh (hunk 3 above)
Comment (danpetry) [reply recipe/build.sh:88]: yes, it was required before but no longer.

File: recipe/meta.yaml (hunk 4 above)
Comment (danpetry) [reply recipe/meta.yaml:119]: Done

File: recipe/meta.yaml (hunk 5 above)
Comment (danpetry) [reply recipe/meta.yaml:120]: done

File: recipe/meta.yaml (hunk 6 above)
Comment (danpetry) [reply recipe/meta.yaml:146]: done

File: recipe/meta.yaml [hunk 7]
```diff
@@ -67,6 +72,8 @@ build:
     - "**/libtorch_cpu.dylib"     # [osx]
     - "**/libtorch_python.dylib"  # [osx]
     - "**/libiomp5.dylib"         # [osx]
+    - "**/libomp.dylib"           # [osx]
```
[the remark this answers is not in the document]
Comment (danpetry) [reply recipe/meta.yaml:75]: added, removed this whitelist entry

File: recipe/meta.yaml [hunk 8]
```diff
@@ -108,13 +114,11 @@ requirements:
...
-    - numpy                           # [    (osx and arm64) or  py>=310]
-    - pip
+    # Pinnings as per pytorch/.circleci/docker/common/install_conda.sh,
+    # except for python 3.8 because we don't have the numpy version required
+    - numpy =1.21.2 # [py>=310]
+    - numpy =1.19.2 # [py<=39]
+    - pip           # Required for in tree builds
     - pkg-config                      # [unix]
     - python
     - pyyaml
```
[the remark this answers is not in the document]
Comment (danpetry) [reply recipe/meta.yaml:124]: removed this range

**Comment by danpetry** [comment]
Note the linter errors are false positives

**Review by skupr-anaconda — APPROVED** [review]

**Review by katietz — APPROVED** [review]
Thanks, was a hard piece of work ...

**Review by katietz — APPROVED** [review]

**Pull request merged**